### DISCUSSION TITLE
Is there a way to prevent certain combinations of operators, like exp(cos(x)) ?

### DISCUSSION THREAD

**Discussion by jburz2001** [MilesCranmer/PySR#69] [Q&A] [answered]
Hello. I'd like to use this package for Symbolic Regression of physics equations. 

Now, physics equations rarely exhibit exponentials of sinusoids. So is there a way for me to tell PySR to not use `exp(cos(x))` ? I know that I can constrain the complexity of `exp` to 1 through `constraints = {'exp':1}`, and that this would remove sinusoids, but I'd like to prevent exp from having the `cos` function as an argument altogether.

Can this be done? If so, how may I do it? I was thinking about defining my own exponential function that cannot take sinusoids as an argument, but I am currently unsure about how to do this.

Thank you!

**Comment by MilesCranmer** [comment]
Hi @jburz2001,

Great question, this is something that I think would be useful for many. You are right that you could disable complex arguments with the `constraints` option, but unfortunately there is no current way to disallow specific patterns in the equation like you are interested in.

While I don't have a timeline for when I would be able to implement this, what I can do is point you to where this could potentially be added if you are interested.

In the backend (https://github.com/MilesCranmer/SymbolicRegression.jl/), there is a function [here](https://github.com/MilesCranmer/SymbolicRegression.jl/blob/856f7b1c80483ec113b4e878119de5ea121f227e/src/CheckConstraints.jl#L49-L68):

```julia
function check_constraints(tree::Node, options::Options, maxsize::Int)::Bool
    if countNodes(tree) > maxsize
        return false
    end
    for i=1:options.nbin
        if options.bin_constraints[i] == (-1, -1)
            continue
        elseif flagBinOperatorComplexity(tree, Val(i), options)
            return false
        end
    end
    for i=1:options.nuna
        if options.una_constraints[i] == -1
            continue
        elseif flagUnaOperatorComplexity(tree, Val(i), options)
            return false
        end
    end
    return true
end
```
This checks the user-passed constraints, and returns `false` if they are violated. It's used during the equation search, so you shouldn't need to change anything else. Here is where you could try to set up a more complex constraint check. 

If interested in trying to implement this:
1. An equation (I call equation variables `tree`, and the equation type `Node`) is a simple binary tree, defined [here](https://github.com/MilesCranmer/SymbolicRegression.jl/blob/856f7b1c80483ec113b4e878119de5ea121f227e/src/Equation.jl). You would want to recursively traverse it to test your constraint. You can see some example functions that operate on equations [here](https://github.com/MilesCranmer/SymbolicRegression.jl/blob/856f7b1c80483ec113b4e878119de5ea121f227e/src/EquationUtils.jl).
2. You can work directly with the Julia backend independently (see the readme), or if you want to test it from Python, you can do: `julia_project="[path/to/SymbolicRegression.jl/repo]"` as an argument to PySR, and it will use the local Julia code instead of the package.


Let me know if you have questions!
Cheers,
Miles

**Reply by jburz2001** [reply]
Thank you for your detailed response, Miles. Apologies for the delay in mine.

I will look into your suggested methods for disallowing specific operator combinations. Thanks again!

**Comment by MilesCranmer** [comment, marked as answer]
Just following up on this. There is now the `nested_constraints` parameter, which is exactly built for this purpose.
